src/data_cell/flatten_datacell.h

```diff
@@ -0,0 +1,194 @@
+
+// Copyright 2024-present the vsag project
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+#include <limits>
+#include <memory>
+
+#include "flatten_interface.h"
+#include "io/basic_io.h"
+#include "quantization/quantizer.h"
+
+namespace vsag {
+
+template <typename QuantTmpl, typename IOTmpl>
+class FlattenDataCell : public FlattenInterface {
+public:
+    FlattenDataCell() = default;
+
+    explicit FlattenDataCell(const std::string& initializeJson);  // todo
+
+    void
+    Query(float* resultDists,
+          std::shared_ptr<ComputerInterface> computer,
+          const uint64_t* idx,
+          uint64_t idCount) override {
+        auto comp = std::static_pointer_cast<Computer<QuantTmpl>>(computer);
+        this->queryImpl(resultDists, comp, idx, idCount);
+    }
+
+    std::shared_ptr<ComputerInterface>
+    FactoryComputer(const float* query) override {
+        return this->factoryComputerImpl(query);
+    }
+
+    float
+    ComputePairVectors(uint64_t id1, uint64_t id2) override;
+
+    void
+    Train(const float* data, uint64_t count) override;
+
+    void
+    InsertVector(const float* vector, uint64_t idx) override;
+
+    void
+    BatchInsertVector(const float* vectors, uint64_t count, uint64_t* idx) override;
+
+    void
+    SetMaxCapacity(uint64_t capacity) override {
+        this->max_capacity_ = std::max(capacity, this->total_count_);  // TODO add warning
+    }
+
+    [[nodiscard]] uint64_t
+    TotalCount() const override {
+        return this->total_count_;
+    }
+
+    [[nodiscard]] const uint8_t*
+    GetCodesById(uint64_t id) const override;
+
+    inline void
+    SetQuantizer(std::shared_ptr<Quantizer<QuantTmpl>> quantizer) {
+        this->quantizer_ = quantizer;
+        this->code_size_ = quantizer_->GetCodeSize();
+    }
+
+    inline void
+    SetIO(std::shared_ptr<BasicIO<IOTmpl>> io) {
+        this->io_ = io;
+    }
+
+public:
+    uint64_t total_count_{0};
+
+    uint64_t max_capacity_{1000000};
+
+    uint64_t code_size_{0};
+
+    std::shared_ptr<Quantizer<QuantTmpl>> quantizer_{nullptr};
+
+    std::shared_ptr<BasicIO<IOTmpl>> io_{nullptr};
```

Comment:
what is the reason to make these variables public? 

remove blank lines, we only add a blank line to distinguish the usage of variables.